libbpf-tools/mountsnoop.bpf.c

```diff
@@ -0,0 +1,131 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/* Copyright (c) 2021 Hengqi Chen */
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "mountsnoop.h"
+
+#define MAX_ENTRIES 10240
+
+const volatile pid_t target_pid = 0;
+static struct event empty = {};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, __u32);
+	__type(value, struct arg);
+} args SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, __u32);
+	__type(value, struct event);
+} values SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(__u32));
+	__uint(value_size, sizeof(__u32));
+} events SEC(".maps");
+
+static int probe_entry(const char *src, const char *dest, const char *fs,
+		       __u64 flags, const char *data, enum op op)
+{
+	__u64 pid_tgid = bpf_get_current_pid_tgid();
+	__u32 pid = pid_tgid >> 32;
+	__u32 tid = (__u32)pid_tgid;
+	struct arg arg = {};
+
+	if (target_pid && target_pid != pid)
+		return 0;
+
+	arg.ts = bpf_ktime_get_ns();
+	arg.flags = flags;
+	arg.src = src;
+	arg.dest = dest;
+	arg.fs = fs;
+	arg.data= data;
+	arg.op = op;
+	bpf_map_update_elem(&args, &tid, &arg, BPF_ANY);
+	bpf_map_update_elem(&values, &tid, &empty, BPF_ANY);
+	return 0;
+};
+
+static int probe_exit(void *ctx, int ret)
+{
+	__u64 pid_tgid = bpf_get_current_pid_tgid();
+	__u32 pid = pid_tgid >> 32;
+	__u32 tid = (__u32)pid_tgid;
+	struct arg *argp;
+	struct event *eventp;
+	struct task_struct *task;
+
+	argp = bpf_map_lookup_elem(&args, &tid);
+	if (!argp)
+		return 0;
+
+	eventp = bpf_map_lookup_elem(&values, &tid);
+	if (!eventp)
+		return 0;
+
+	task = (struct task_struct *)bpf_get_current_task();
+	eventp->delta = bpf_ktime_get_ns() - argp->ts;
+	eventp->flags = argp->flags;
+	eventp->pid = pid;
+	eventp->tid = tid;
+	eventp->mnt_ns = BPF_CORE_READ(task, nsproxy, mnt_ns, ns.inum);
+	eventp->ret = ret;
+	eventp->op = argp->op;
+	bpf_get_current_comm(&eventp->comm, sizeof(eventp->comm));
+	if (argp->src)
+		bpf_probe_read_user_str(eventp->src, sizeof(eventp->src), argp->src);
+	if (argp->dest)
+		bpf_probe_read_user_str(eventp->dest, sizeof(eventp->dest), argp->dest);
+	if (argp->fs)
+		bpf_probe_read_user_str(eventp->fs, sizeof(eventp->fs), argp->fs);
+	if (argp->data)
+		bpf_probe_read_user_str(eventp->data, sizeof(eventp->data), argp->data);
+	bpf_perf_event_output(ctx, &events, BPF_F_CURRENT_CPU, eventp, sizeof(*eventp));
+
+	bpf_map_delete_elem(&args, &tid);
+	bpf_map_delete_elem(&values, &tid);
+	return 0;
+}
+
+SEC("tracepoint/syscalls/sys_enter_mount")
+int mount_entry(struct trace_event_raw_sys_enter *ctx)
```

Comment:
not sure if you realized, but you could have written this as 

```
int BPF_PROG(mount_entry, const char *src, const char *dst, const char *fs, __u64 flags, const char data) {
    return probe_entry(src, dest, fs, flags, data, MOUNT);
}
```

Reply:
Just like the way we use `fentry`, Right? 
Actually, I didn't realize this. Thanks for the reminder.
Maybe we can add an example to libbpf-bootstrap.

Reply:
`mount` has the following format, so the first argument has to be an int.
```
field:int __syscall_nr;	offset:8;	size:4;	signed:1;
field:char * dev_name;	offset:16;	size:8;	signed:0;
field:char * dir_name;	offset:24;	size:8;	signed:0;
field:char * type;	offset:32;	size:8;	signed:0;
field:unsigned long flags;	offset:40;	size:8;	signed:0;
field:void * data;	offset:48;	size:8;	signed:0;
```

Reply:
I will do this for the `kvmexit` tool.

Reply:
that format is for non-raw tracepoint. BPF_PROG() is just rewriting those ctx[0], ctx[1] accesses into a nice input parameters. So don't overthink this.

Reply:
@anakryiko  I got this error when I tried this method.

```
libbpf: load bpf program failed: Permission denied
libbpf: -- BEGIN DUMP LOG ---
libbpf: 
arg#0 type is not a struct
Unrecognized arg#0 type PTR
; int BPF_PROG(mount_entry, const char *src, const char *dest, const char *fs, __u64 flags, const char *data)
0: (79) r2 = *(u64 *)(r1 +32)
1: (7b) *(u64 *)(r10 -80) = r2
2: (79) r2 = *(u64 *)(r1 +24)
3: (7b) *(u64 *)(r10 -72) = r2
4: (79) r6 = *(u64 *)(r1 +16)
5: (79) r9 = *(u64 *)(r1 +8)
6: (79) r8 = *(u64 *)(r1 +0)
invalid bpf_context access off=0 size=8
processed 7 insns (limit 1000000) max_states_per_insn 0 total_states 0 peak_states 0 mark_read 0

libbpf: -- END LOG --
libbpf: failed to load program 'mount_entry'
libbpf: failed to load object 'mountsnoop_bpf'
libbpf: failed to load BPF skeleton 'mountsnoop_bpf': -4007
failed to load BPF object: -4007
```

Reply:
Oh, wait. I got confused, sorry. I somehow decided that you are using *raw tracepoint* (`SEC("raw_tp/sys_enter")`) but clearly you are not. BPF_PROG() approach I described above is applicable only to raw_tracepoints, so never mind.